### PR TITLE
[agent-c][issue #1535] Add guard escalation payload fields

### LINKED ISSUE #1535: [design/runtime] Add explicit guard escalation failure-context payload authoring
## Context

Split from #1462 Point 14.

Point 14 contained two different concerns:

1. `on_fail: kill` moves the entity to a terminal/absorbing state and cannot be reopened.
2. `on_fail: escalate:{event}` emits an escalation event with an empty explicit business payload by default.

The first concern is accepted platform semantics. `platform-spec.yaml` says terminal states are absorbing, no event can reopen them, and reopening requires creating a new entity. That should not be reopened here.

The second concern is a real authoring friction: guard escalation is often where authors want failure context, but today `on_fail: escalate:{event}` has no payload authoring surface.

## Current behavior

Runtime intentionally does not implicitly pass through trigger payload or entity metadata on guard escalation.

Implementation evidence:

- `internal/runtime/engine/executor.go` handles `GuardFailureEscalate` by calling the emit payload shaper with an empty explicit business payload.
- `internal/runtime/engine/executor_test.go` asserts guard escalation payload is empty.
- `internal/runtime/pipeline/handler_engine_transaction_test.go` asserts guard escalation does not leak trigger payload, legacy payload, or entity metadata.
- `internal/runtime/bootverify/workflow_payload_completeness_checks.go` treats `guard.on_fail.escalate` as an emit site and fails if the target event requires business fields that are not statically provable.

This is safe, but not very ergonomic: the escalation handler loses the guard failure context unless the event schema is empty or envelope-only.

## Design question

Should guard escalation gain an explicit payload/failure-context authoring surface?

Possible designs to gate:

- Structured object form:

```yaml
guard:
  id: score_check
  check: payload.score >= policy.threshold
  on_fail:
    escalate:
      event: check.escalated
      fields:
        score: payload.score
        threshold: policy.threshold
        reason: "score_below_threshold"
```

- Dedicated failure context fields exposed to the escalation payload shaper, for example `guard.id`, `guard.failed_check`, `guard.reason_code`, with explicit mapping required.
- Keep the current string shorthand `on_fail: "escalate:event.name"` as envelope-only and add the object form only when payload is needed.
- Leave current behavior unchanged and document that guard escalation must target an empty/envelope-only event.

## Constraints

- No implicit passthrough from trigger payload.
- No implicit passthrough from entity metadata.
- No envelope-owned fields in emitted business payload.
- Keep runtime-owned event envelope/context construction authoritative.
- Do not change `on_fail: kill`; terminal irreversibility is accepted semantics.
- Any implementation PR must update `platform-spec.yaml` and code/conformance together.

## Proof expectations if implemented

A focused PR should prove:

- Existing string shorthand remains envelope-only or is intentionally migrated by spec.
- Explicit guard escalation fields can satisfy required business payload fields.
- Missing required escalation fields fail verify.
- Trigger payload/entity metadata do not leak unless explicitly mapped from an allowed expression root.
- Envelope-owned fields remain rejected.

Related: #1462

### PR DESCRIPTION
**Pull request by yazzaoui**
## Summary

- Add a normalized guard failure/escalation model that preserves scalar `on_fail: "escalate:event"` as envelope-only and adds structured `on_fail.escalate.event/fields` for explicit business payload authoring.
- Wire the same canonical guard escalation `EmitSpec` through runtime execution, bootverify expression/payload checks, semanticview authored emit sites, and served pipeline proof.
- Update `platform-spec.yaml` to define object-form guard escalation payload semantics and guard-time field evaluation.

Closes #1535

## Governing Refs / Context

- `platform-spec.yaml` handler guard `on_fail` schema.
- `platform-spec.yaml` `handler_outcome.escalate` semantics.
- #1535 issue body and approved `Pre-Implementation Coverage Audit`.
- #1462 Point 14 split/tracker context.
- Gate outcome on #1535: `approved as first slice` for `guard_escalation_explicit_payload_authoring_gap`.

Requested `docs/IMPLEMENTER_GUIDELINES.md` and `docs/SEMANTIC_DRIFT.md` are not present on current `origin/master`; the issue/spec/code contracts above are the binding context.

## Semantic Migration

- New canonical owner: `runtimecontracts.GuardFailureSpec` / `GuardEscalationSpec`-equivalent normalized model via `GuardSpec.FailureSpec()`, carrying a guard escalation `EmitSpec` with `event` and `fields`.
- Old producer/reader/writer path now invalid: treating `GuardSpec.OnFail string` plus local `ParseGuardFailure` as the complete guard escalation model. Scalar `OnFail` remains a shorthand carrier only.
- Production paths checked/moved: YAML contract decode, runtime `Executor.applyGuardFailure`, bootverify guard action validation, bootverify payload completeness, bootverify expression/entity lifecycle checks, semanticview authored emit sites, served pipeline execution.
- Migration completeness: complete for #1535 object-form explicit payload authoring. `guard.*` failure-detail roots remain intentionally split unless a later gate widens scope.

## Proof

- `go test ./internal/runtime/contracts ./internal/runtime/engine ./internal/runtime/bootverify ./internal/runtime/semanticview ./internal/runtime/pipeline -run 'Guard|OnFail|Escalate|AuthoredEmitSites|PayloadCompleteness|EntityFieldsAvailableBeforeGuardEscalation' -count=1`
- `go test ./internal/runtime/bootverify -run 'Guard|PayloadCompleteness' -count=1`
- `go test ./...`

### CODE REVIEW

**Comment by yazzaoui** [comment]
## Post-Implementation Proof Audit

Exact semantic concepts changed:
- Guard escalation payload authoring: `on_fail.escalate.event/fields` can now author explicit business payload fields.
- Guard escalation canonical owner: guard failure semantics now normalize through `runtimecontracts.GuardFailureSpec` via `GuardSpec.FailureSpec()` instead of treating `GuardSpec.OnFail string` as the complete model.
- Guard escalation emit-site consumption: runtime, bootverify, and semanticview now consume the same guard escalation `EmitSpec`.
- Guard escalation field lifecycle: explicit fields evaluate at guard-failure time with guard/pre-handler visibility, not ordinary late `emit.fields` visibility.

Chosen working failure class closed entirely: `guard_escalation_explicit_payload_authoring_gap`.

Parent failure class above it: declarative failure-emission payload ownership/validation across spec, contracts, verify, semanticview, and runtime.

Canonical owners after the change:
- Spec: `platform-spec.yaml` guard `on_fail` schema and `handler_outcome.escalate` semantics.
- Contract/YAML: `runtimecontracts.GuardFailureSpec`, `GuardSpec.FailureSpec()`, and guarded YAML decode for scalar shorthand vs object form.
- Runtime: `Executor.applyGuardFailure` consuming `GuardSpec.FailureSpec()` and evaluating the escalation `EmitSpec` fields before shaping/queueing.
- Verification: bootverify guard action validation, guard escalation expression validation, entity lifecycle validation, and payload completeness using the normalized guard escalation `EmitSpec`.
- Semantic inventory: `semanticview.AuthoredEmitSites` exposing guard escalation event+fields from the normalized owner.
- Served/runtime proof: pipeline execution tests proving emitted payload and no-leak behavior.

Touched consumers of each owner:
- Contract/YAML owner consumers: contract YAML decode tests, runtime executor, bootverify validation/completeness, semanticview authored emit-site collection, pipeline test setup.
- Runtime owner consumers: engine guard execution, payload shaper, emit intent queueing, served pipeline coordinator.
- Bootverify owner consumers: condition validation, emit field expression validation, expression entity lifecycle checks, semantic drift payload completeness, pin/output-pin consumers via semanticview.
- Semanticview owner consumers: authored emit-site tests and downstream pin/output-pin checks that consume `AuthoredEmitSites` rather than parsing guard strings.

Exhaustive systematic-consumption audit:
- `GuardSpec.OnFail string`: still survives only as scalar shorthand carrier and display/test compatibility path. It is no longer authoritative for object-form payload semantics.
- `runtimecontracts.GuardFailureSpec`: new canonical structured owner for action + escalation `EmitSpec`.
- `runtimeengine.ParseGuardFailure`: preserved as scalar compatibility wrapper over contract parsing; not the object-form owner.
- `Executor.applyGuardFailure`: moved to `GuardSpec.FailureSpec()` and explicit field evaluation.
- `workflow_expression_checks.go`: moved guard action validation and guard escalation fields to normalized spec; removed obsolete string helper.
- `workflow_payload_completeness_checks.go`: moved from synthetic event-only `EmitSpec` to normalized event+fields.
- `semanticview/authored_emit_sites.go`: moved from string parsing to normalized event+fields.
- Pin routing/output pin checks: already consume semanticview and therefore consume the corrected path without independent parser changes.
- Ordinary `handler.emit`, rules, `on_complete`, `accumulate.on_timeout`, `fan_out.emit`: checked as different concepts already using `EmitSpec`; unchanged.
- `reject`, `discard`, `kill`: checked as different guard failure concepts; unchanged.
- Agent/action/artifact/mailbox payload surfaces: checked as different payload concepts; unchanged.

Old paths now invalid / non-authoritative / still surviving:
- Invalid as complete owner: `GuardSpec.OnFail string` plus local scalar parser as the full guard escalation model.
- Non-authoritative but surviving: scalar `on_fail: "escalate:event"` shorthand, intentionally envelope-only.
- Removed/invalidated: bootverify and semanticview event-only synthesis for object-form guard escalation.
- Still surviving by design: scalar shorthand envelope-only behavior and `kill` terminal behavior.

Broader failure class: declarative emit-like failure surfaces need one payload ownership model across spec, contracts, runtime, verify, and indexing.

Was the issue symptom-shaped: yes, partly. The symptom was empty guard escalation payload. The implementation treated that line as an entry point and moved the full same-concept consumer set.

Achieved closure level: complete closure for #1535's approved first slice: structured object-form explicit guard escalation payload authoring. This PR should close #1535.

Parent class remains open: yes. The broader parent remains open for any future failure-emission payload ownership work outside guard escalation.

Remaining child-slice tail estimate for parent: one optional medium slice if product wants `guard.*` failure-detail roots (`guard.failed_check`, `guard.reason_code`). Implementation did not change the estimate; confidence remains medium. Not required for #1535 closure.

Sibling contexts checked:
- `on_fail: kill`, `reject`, `discard`: split/different concepts, unchanged.
- Ordinary declarative `emit.fields`: existing canonical payload owner, unchanged.
- Rule/on_complete/accumulate/fan_out emit sites: existing `EmitSpec` consumers, unchanged.
- Pin routing/output pin checks: consume semanticview corrected owner, no independent same-concept parser found.
- Action/artifact/mailbox/agent emits: different payload surfaces, unchanged.

Generic failing proof used or created:
- Created object-form bootverify proof that required fields are satisfied.
- Created object-form bootverify proof that missing required fields still fail.
- Created object-form bootverify proof that envelope-owned payload fields are rejected.
- Created missing-event validation proof.
- Created runtime and served pipeline proof that explicit fields are emitted and unmapped trigger/entity metadata do not leak.
- Created semanticview proof that authored guard escalation carries fields.
- Created lifecycle proof that guard escalation fields have guard-time visibility.

Watchlist decision: keep mapped to #1535 under #1462 Point 14. No repo-local watchlist file exists for this issue; unrelated watchlist entries were not updated. No new issue required unless `guard.*` failure-detail roots are pursued.

Deeper architecture/type-model smell: `guard.on_fail` had become stringly typed while the platform already had structured `EmitSpec` for authored payload surfaces. Long-run direction is to keep guard failure semantics normalized and avoid reintroducing independent string parsers.

Tracking decision for architecture feedback: tracked in #1535 for this object-form normalization. Optional `guard.*` failure detail roots should be a child/gate repair if requested later.

Effort/ROI estimate for architecture smell: this PR handles the high-ROI slice. Further broad guard failure cleanup beyond `GuardFailureSpec` would be low-to-medium ROI unless new failure-context roots are needed.

Manifestation coverage:

| Manifestation | Coverage status | Exact proof |
| --- | --- | --- |
| Scalar `on_fail: "escalate:event"` emits empty business payload | execution-proven through the same corrected path | `TestExecutor_GuardOnFailEscalateCreatesEmitIntent`; existing served no-leak test |
| Structured object fields satisfy required escalation payload | reproduced and fixed | `TestRun_DoesNotWarnWhenGuardEscalateObjectFieldsCoverRequiredPayload`; `TestExecutor_GuardOnFailEscalateObjectFieldsShapeExplicitPayload`; `TestExecuteNodeContractHandler_GuardEscalateObjectFieldsUseExplicitPayloadOnly` |
| Missing required object-form escalation field fails verify | reproduced and fixed | `TestRun_ErrorsWhenGuardEscalateObjectFieldsMissRequiredPayload` |
| Trigger payload/entity metadata do not leak implicitly | execution-proven through the same corrected path | `TestExecuteNodeContractHandler_GuardEscalateUsesOnlyRuntimeOwnedEnvelope`; `TestExecuteNodeContractHandler_GuardEscalateObjectFieldsUseExplicitPayloadOnly` |
| Envelope-owned fields rejected in escalation business payload | reproduced and fixed | `TestRun_ErrorsWhenGuardEscalateObjectFieldsAuthorEnvelopeOwnedField` |
| Missing escalation event rejected | reproduced and fixed | `TestRun_RejectsGuardEscalateObjectMissingEvent` |
| Invalid nested scalar object shortcut rejected | reproduced and fixed | `TestGuardSpecDecode_RejectsNestedScalarEscalateShortcut` |
| Semanticview/pin accounting sees object-form fields | reproduced and fixed | `TestAuthoredEmitSites_GuardEscalationObjectCarriesFields` |
| Guard escalation fields use guard-time visibility | reproduced and fixed | `TestWorkflowEntityFieldsAvailableBeforeGuardEscalation_UsesGuardTimeVisibility` |
| `guard.*` failure-detail namespace | split / escalated as separate class | Explicitly excluded by approved gate; no code added |
| `on_fail: kill` terminal behavior | split / escalated as separate class | Explicitly accepted by #1462/#1535; unchanged |

Required supported-surface proof actually run:
- `go test ./internal/runtime/contracts ./internal/runtime/engine ./internal/runtime/bootverify ./internal/runtime/semanticview ./internal/runtime/pipeline -run 'Guard|OnFail|Escalate|AuthoredEmitSites|PayloadCompleteness|EntityFieldsAvailableBeforeGuardEscalation' -count=1`
- `go test ./internal/runtime/bootverify -run 'Guard|PayloadCompleteness' -count=1`
- `go test ./...`

Optional general feedback: no extra broad refactor was needed; the one meaningful architecture improvement was making guard failure normalization explicit enough that all same-concept consumers can share it.

File: internal/runtime/contracts/workflow_contract_yaml_rules.go [hunk 1]
```diff
@@ -21,18 +21,110 @@ func (g *GuardSpec) UnmarshalYAML(node *yaml.Node) error {
...
+	switch node.Kind {
+	case yaml.ScalarNode:
+		value := strings.TrimSpace(node.Value)
+		if value == "" {
+			return "", GuardFailureSpec{}, nil
+		}
+		return value, GuardFailureSpec{}, nil
+	case yaml.MappingNode:
+		if len(node.Content) == 0 {
+			return "", GuardFailureSpec{}, nil
```
Comment (yazzaoui) [inline internal/runtime/contracts/workflow_contract_yaml_rules.go]: This treats an explicit object-form `on_fail: {}` exactly like an omitted `on_fail`. That silently falls through to default `reject` via `GuardSpec.FailureSpec()`, but the new spec only allows scalar actions or `{escalate: {event, fields}}`, and the pre-audit/proof audit require invalid object forms to fail closed. Please reject an empty mapping, and any mapping without `escalate`, and add a regression test for the explicit empty-object case.

**Review by yazzaoui — COMMENTED** [review]
Changes needed.

Blocking finding:
- `internal/runtime/contracts/workflow_contract_yaml_rules.go` accepts an explicit empty `guard.on_fail` mapping as if `on_fail` were absent. Because `GuardSpec.FailureSpec()` then parses the empty scalar path as default `reject`, `on_fail: {}` silently becomes a valid reject guard instead of failing closed. The governing spec now only permits scalar actions or `{escalate: {event, fields}}`, and the #1535 pre-audit/proof audit explicitly require invalid object forms and missing escalation events to be rejected. This leaves one object-form validation manifestation unclosed. Please reject an explicit empty mapping, and any mapping without the supported `escalate` key, during YAML/contract decode or equivalent contract validation, and add a regression test.

Process assessment:
- Issue category is correct: design/runtime + spec/conformance authoring-surface gap.
- The chosen failure class, `guard_escalation_explicit_payload_authoring_gap`, is acceptable as the approved first slice under #1462 Point 14; it is not broad parent closure.
- The branch updates the authoritative `platform-spec.yaml` and moves the main same-concept consumers together: contract normalization, runtime execution, bootverify expression/payload checks, semanticview authored emit-site accounting, and served pipeline proof.
- I did not find a hidden runtime/verify/semanticview same-concept interpreter still bypassing the normalized guard escalation owner. Catalog runner references are non-authoritative test harnesses.
- Tracker/watchlist state remains #1535 under #1462 Point 14. The repo has no `docs/POTENTIAL_ISSUES.md`; no new tracker issue is needed for this narrow decoder fix.

Merge recommendation: do not merge until the malformed explicit object-form path is fail-closed and covered by a contract/YAML regression test.

**Comment by yazzaoui** [comment]
Review checklist for PR #1581:

- Guideline/spec checks: failed. `on_fail: {}` is an explicit malformed object form and currently decodes as default `reject` instead of failing closed.
- Risk level: medium semantic/spec-validation risk; main runtime payload path looks coherent, but invalid object-form validation is incomplete.
- Closure level: not mergeable; #1535 first-slice closure is blocked on one fail-closed decoder manifestation.
- Residual risk after fix: low if the decoder rejects empty/missing-top-level object form and a contract/YAML regression test is added.
- Tests run locally: `go test ./internal/runtime/contracts ./internal/runtime/engine ./internal/runtime/bootverify ./internal/runtime/semanticview ./internal/runtime/pipeline -run 'Guard|OnFail|Escalate|AuthoredEmitSites|PayloadCompleteness|EntityFieldsAvailableBeforeGuardEscalation' -count=1`; `go test ./internal/runtime/bootverify -run 'Guard|PayloadCompleteness' -count=1`; `go test ./...`.
- Tracker/watchlist: no new issue; keep mapped to #1535 under #1462 Point 14. No repo-local `docs/POTENTIAL_ISSUES.md` exists.
- Merge recommendation: changes needed; do not merge until malformed explicit object-form guard `on_fail` fails closed.

File: internal/runtime/contracts/workflow_contract_yaml_rules.go (hunk 1 above)
Comment (yazzaoui) [reply internal/runtime/contracts/workflow_contract_yaml_rules.go]: Addressed in `8929be69`.

Changes:
- `guard.on_fail: {}` now fails closed during YAML decode instead of falling through to default `reject`.
- object-form `guard.on_fail` without the supported `escalate` key remains rejected.
- added `TestGuardSpecDecode_RejectsMalformedOnFailObjectForms` covering both explicit empty object and missing-`escalate` object forms.

Proof run:
- `go test ./internal/runtime/contracts -run 'GuardSpecDecode' -count=1`
- `go test ./internal/runtime/contracts ./internal/runtime/engine ./internal/runtime/bootverify ./internal/runtime/semanticview ./internal/runtime/pipeline -run 'Guard|OnFail|Escalate|AuthoredEmitSites|PayloadCompleteness|EntityFieldsAvailableBeforeGuardEscalation' -count=1`
- `go test ./...`

**Comment by yazzaoui** [comment]
## Post-Implementation Proof Audit Addendum

Review fix commit: `8929be69`.

The review finding was valid: explicit malformed object-form `guard.on_fail: {}` was still accepted as omitted `on_fail`, then defaulted to `reject`. That was inside the approved #1535 class because object-form guard escalation authoring must fail closed when malformed.

Semantic concept, canonical owners, parent class, and closure commitment are unchanged from the posted proof audit. This addendum records the newly covered malformed-object manifestation.

Additional manifestation coverage:

| Manifestation | Coverage status | Exact proof |
| --- | --- | --- |
| Explicit empty object-form `guard.on_fail: {}` silently defaulted to `reject` | reproduced and fixed | `TestGuardSpecDecode_RejectsMalformedOnFailObjectForms/empty_object` |
| Object-form `guard.on_fail` without supported `escalate` key accepted or normalized | reproduced and fixed | `TestGuardSpecDecode_RejectsMalformedOnFailObjectForms/missing_escalate_key` |

Required proof rerun:
- `go test ./internal/runtime/contracts -run 'GuardSpecDecode' -count=1`
- `go test ./internal/runtime/contracts ./internal/runtime/engine ./internal/runtime/bootverify ./internal/runtime/semanticview ./internal/runtime/pipeline -run 'Guard|OnFail|Escalate|AuthoredEmitSites|PayloadCompleteness|EntityFieldsAvailableBeforeGuardEscalation' -count=1`
- `go test ./...`

Closure update: #1535 first-slice closure is now restored for malformed object-form validation as well as runtime/verify/semanticview payload consumption.

**Review by yazzaoui — COMMENTED** [review]
Approved by process; GitHub would not allow a formal `APPROVE` review from the PR owner account.

I re-reviewed head `8929be69` after the decoder fix. The previous blocking finding is resolved: `decodeGuardOnFailNode` now rejects explicit empty `guard.on_fail` mappings instead of treating them as omitted/default `reject`, and `TestGuardSpecDecode_RejectsMalformedOnFailObjectForms` covers both `on_fail: {}` and object-form mappings without the supported `escalate` key.

Spec/process assessment:
- Governing spec was updated in `platform-spec.yaml` for object-form guard escalation fields and guard-time evaluation.
- Issue category remains correct: design/runtime + spec/conformance authoring-surface gap.
- Failure-class framing remains acceptable as the approved first slice: `guard_escalation_explicit_payload_authoring_gap` under #1462 Point 14.
- The PR now moves the same-concept consumers together: YAML/contract normalization, runtime execution, bootverify action/expression/payload checks, semanticview authored emit-site accounting, lifecycle visibility, and served pipeline proof.
- I did not find a live hidden same-concept interpreter after the fix. Remaining `OnFail` string use is scalar shorthand/test harness compatibility, not the object-form payload owner.
- Parent class remains tracked by #1462; optional `guard.*` failure-detail roots remain split and are not required for #1535 closure.
- The previous unresolved thread was verified obsolete on current head and resolved.

Closure assessment: #1535 first-slice failure class is eliminated. No worthwhile follow-up remains for this PR scope. Merge recommendation: approve/merge.

**Comment by yazzaoui** [comment]
Review checklist for PR #1581 re-review:

- Guideline/spec checks: pass. Authoritative `platform-spec.yaml` is updated with object-form guard escalation payload semantics.
- Risk level: low after fix; malformed object-form `guard.on_fail` now fails closed.
- Closure level: #1535 first-slice failure class eliminated for `guard_escalation_explicit_payload_authoring_gap`.
- Residual risk: optional `guard.*` failure-detail roots remain intentionally split under #1535/#1462 context and are not closure-bearing for this PR.
- Tests run locally: `go test ./internal/runtime/contracts -run 'GuardSpecDecode' -count=1`; `go test ./internal/runtime/contracts ./internal/runtime/engine ./internal/runtime/bootverify ./internal/runtime/semanticview ./internal/runtime/pipeline -run 'Guard|OnFail|Escalate|AuthoredEmitSites|PayloadCompleteness|EntityFieldsAvailableBeforeGuardEscalation' -count=1`; `go test ./...`.
- Watchlist/tracker: no repo-local `docs/POTENTIAL_ISSUES.md`; keep mapped to #1535 under #1462 Point 14. No new issue needed.
- Merge recommendation: approve/merge. Formal GitHub approval was unavailable because the authenticated account owns the PR.
